weave/trace/refs.py

```diff
@@ -15,16 +15,33 @@ class WeaveDigestError(ValueError):
     """Raised when a digest is invalid."""
 
 
-@dataclasses.dataclass(frozen=True)
+@dataclass(frozen=True)
 class Ref:
     def uri(self) -> str:
         raise NotImplementedError
 
     def as_param_dict(self) -> dict:
-        return dataclasses.asdict(self)
+        return asdict(self)
 
+    def __deepcopy__(self, memo: dict) -> "Ref":
+        d = {}
+        for f in fields(self):
+            v = getattr(self, f.name)
 
-@dataclasses.dataclass(frozen=True)
+            # Must resolve the future to remove the lock which causes issues when deepcopying
+            if isinstance(v, Future):
+                v = v.result()
```

Comment:
you could just copy the future, no? Else this internal deep copy with all of a sudden block the user process

Reply:
yeah good point -- I thought you had to resolve, but shallow copy is simpler and better